dbms/src/Interpreters/HashJoinSpillContext.cpp

```diff
@@ -27,29 +27,37 @@ HashJoinSpillContext::HashJoinSpillContext(
     , max_cached_bytes(std::max(
           build_spill_config.max_cached_data_bytes_in_spiller,
           probe_spill_config.max_cached_data_bytes_in_spiller))
-{
-    /// join does not support auto spill mode
-    auto_spill_mode = false;
-}
+{}
 
 void HashJoinSpillContext::init(size_t partition_num)
 {
     partition_revocable_memories = std::make_unique<std::vector<std::atomic<Int64>>>(partition_num);
     partition_is_spilled = std::make_unique<std::vector<std::atomic<bool>>>(partition_num);
+    partition_spill_status = std::make_unique<std::vector<std::atomic<AutoSpillStatus>>>(partition_num);
     for (auto & memory : *partition_revocable_memories)
         memory = 0;
-    for (auto & status : *partition_is_spilled)
-        status = false;
+    for (auto & is_spilled : *partition_is_spilled)
+        is_spilled = false;
+    for (auto & spill_status : *partition_spill_status)
+        spill_status = AutoSpillStatus::NO_NEED_AUTO_SPILL;
 }
 
 Int64 HashJoinSpillContext::getTotalRevocableMemoryImpl()
 {
     Int64 ret = 0;
-    for (const auto & x : *partition_revocable_memories)
-        ret += x;
+    for (size_t part_index = 0; part_index < partition_revocable_memories->size(); ++part_index)
+    {
+        if (in_build_stage || isPartitionSpilled(part_index))
+            ret += (*partition_revocable_memories)[part_index];
+    }
     return ret;
 }
 
+bool HashJoinSpillContext::supportFurtherSpill() const
+{
+    return in_spillable_stage && (in_build_stage || isSpilled());
+}
```

Comment:
why not calling `finishSpillableStage()` in `workAfterBuildFinish`.

Reply:
Maybe rename `finishSpillableStage` to `finishTriggerableStage`

Reply:
Then maybe `in_build_stage` doesn't need

Reply:
1. Can't call `finishSpillableStage()` in `workAfterBuildFinish` because once `finishSpillableStage()` is called, the auto spill trigger will think the current operator does not support spill anymore, this is not true for hash join here because after build, the probe side still can(or must) spill if current partition is spilled during build stage.
2. The spill ability for hash join build stage and hash join probe stage is different:
- In hash join build stage, it can spill a new partition
- In hash join probe stage, only the partition that is already spilled during build stage can support spill
so, there must be a flag to let auto spill trigger know the stage, so can not  remove `in_build_stage`